fixes issue #17, implemented revoke

raffis opened the pull request.

tikurahul commented:
Thanks for your PR. Will take a look at this soon.

tikurahul commented:
Hi @raffis. Thanks again. While you are waiting for your PR to be reviewed, can you please go ahead and sign the CLA ? (https://github.com/openid/AppAuth-JS/blob/master/CONTRIBUTING.md#execute-the-contributor-license-agreement-cla)

raffis commented:
quite complicated to support you guys :) ...okay I've sent the CLA.

mcastany commented on src/token_request_handler.ts:
```diff
@@ -46,6 +51,24 @@ export class BaseTokenRequestHandler implements TokenRequestHandler {
+    let revokeTokenResponse = this.requestor.xhr<any>({
+      url: configuration.revocationEndpoint + '?' + this.utils.stringify(request.toStringMap()),
+      dataType: 'json',  // adding implicit dataType
+      method: 'GET'
```
shouldn't this be a POST and the token be sent in the body ? - https://tools.ietf.org/html/rfc7009#section-2.1

tikurahul replied:
Good point. It should 👍

raffis replied:
good point, will fix that as well.
And we need check the response code only: (instead and empty object {}):
> The content of the response body is ignored by the client as all
>   necessary information is conveyed in the response code.

probably I need to change xhr as well to return the response code, and I will extend the revoke class to include param token_type_hint besides token.

raffis replied:
And:
>  The client also includes its authentication credentials as described
>  in Section 2.3. of [RFC6749].

Not sure about that, for example the google idp does not require this, I was able to revoke token without an authorization header, gonna check what happens if I send it along anyway to be correct according rfc.

jhiner replied:
The fact is that Google's implementation appears to require GET on the revocation endpoint, instead of POST https://developers.google.com/identity/protocols/OAuth2WebServer#tokenrevoke. The client authentication is also an open question. In order to work with most spec-conformant providers, I think this library should support client authentication.

jhiner replied:
+1 for `token_type_hint` support

tikurahul replied:
You should be able to reject the promise when an `xhr` fails with the appropriate `AppAuthError` with the underlying HTTP error code. If `xhr` does not do that already, it should.

tikurahul commented:
Apologies, this is getting delayed. I am in India for a family emergency. I will get back to this PR after the 9th of September.

raffis commented:
thats fine, I did not find the time yet to append those fixes above anyway.

raffis commented:
@tikurahul just found out that a revocation endpoint is not even part of the discovery document: http://openid.net/specs/openid-connect-discovery-1_0.html Google does include a revocation_endpoint (We do as well), but do we want implement such a thing if it is not mentioned there? I actually do not get why such an important endpoint is not part of the standard.

raffis commented:
well there is a draft which specifies it: https://tools.ietf.org/html/draft-ietf-oauth-discovery-06

tikurahul commented:
Note to self: There is more information about the revocation process itself in this [RFC](https://tools.ietf.org/html/rfc7009#section-2.1).

raffis commented:
@tikurahul couple things:
* changed revoke to POST
* added support for token_type_hint

* regarding response code on node: 
https://github.com/raffis/AppAuth-JS/blob/master/src/node_support/node_requestor.ts#L35
I could not reject the promise since error here is always null, only if the request itself failed "error" wont be null. it does now reject if it is not http 200, maybe you should look into that.

* revocation request is now authenticated as well, the https://tools.ietf.org/html/rfc7009#section-2.1 actually says that this request must be authenticated if client credentials given. it will be sent as client_secret_post (http://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication)
(For example google authorization server does not like that, it will throw an error if credentials will be sent along, but since they are optional for the revocation request it should be fine, for example our auth server actually requires client authentication for that endpoint according to rfc7009)

* another new issue would be to actually support client_secret_basic (http://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication) besides client_secret_post, and this applies to both the token endpoint and the revocation endpoint

tikurahul commented on src/revoke_token_request.ts:
```diff
@@ -0,0 +1,85 @@
+    return this.clientId !== undefined && this.clientSecret !== undefined;
+  }
+
+  getBasicAuthorizationHeader(): string {
```
Remove unused methods, as we don't have full support for Basic Auth at the moment.

tikurahul commented on src/token_request_handler.ts:
```diff
@@ -46,6 +51,33 @@ export class BaseTokenRequestHandler implements TokenRequestHandler {
+      request: RevokeTokenRequest): Promise<boolean> {
+    let headers: StringMap = {'Content-Type': 'application/x-www-form-urlencoded'};
+
+    // TODO: client_secret_basic support
```
Same here. Remove comments and methods you are not using.

tikurahul commented on src/token_request_handler.ts:
```diff
@@ -56,14 +88,18 @@ export class BaseTokenRequestHandler implements TokenRequestHandler {
+                new AppAuthError(response.error, TokenError.fromJson(response)));
+          }
+        })
+        .catch(response => {
```
Why are you re-catching and re-throwing it as another AppAuth error. Seems unnecessary.

tikurahul commented on src/revoke_token_request.ts:
```diff
@@ -0,0 +1,85 @@
+export class RevokeTokenRequest {
+  constructor(
+      public token: string,
+      public tokenTypeHint?: string,
```
So `tokenTypeHint` can be one of `access_token` and `refresh_token`. So rather than leave this as a `string` define a new type. Like so:

```typescript
type TokenTypeHint = 'refresh_token' | 'access_token';
```

Then the tokenTypeHint can use this type.

tikurahul commented:
Can you also please rebase your changes with `openid/master` that way we can avoid spurious merge commits.

raffis commented:
@tikurahul thanks for your inputs, typescript is not (yet) my native tongue :P, I have added your suggestions.

tikurahul commented on src/revoke_token_request.ts:
```diff
@@ -0,0 +1,78 @@
+    };
+  }
+
+  toStringMap(): StringMap {
```
`toStringMap()` and `toJson()` are very similar in what they do. I know this is meant to coerce all subtypes so you can convert this to a queryString, however, it feels weird. Maybe we should consistently not emit fields that are undefined or null in toJson() and avoid toStringMap() altogether.

tikurahul commented on src/token_request_handler.ts:
```diff
@@ -46,6 +51,26 @@ export class BaseTokenRequestHandler implements TokenRequestHandler {
+        .then(response => {
+          return true;
+        })
+        .catch(error => {
```
You don't need to catch and rethrow. Just remove the catch block. The promise chain will get rejected anyway.

raffis commented:
Sounds good, so do you want me to remove toStringMap() ?, aight did that and toJson() now only includes set values

tikurahul commented on src/token_request_handler.ts:
```diff
@@ -46,6 +51,22 @@ export class BaseTokenRequestHandler implements TokenRequestHandler {
+      url: configuration.revocationEndpoint,
+      method: 'POST',
+      dataType: 'json',  // adding implicit dataType
+      headers: {'Content-Type': 'application/x-www-form-urlencoded'},
```
Nitpick: Can you check if the `Content-Type` is set on the request as expected ? I am confused because the data is not being sent as a bunch of URL encoded parameters. It is being sent as JSON and `dataType` `json` will actually set the `Content-Type` to the `json` mime type. In which case you can remove this line in a subsequent PR.

tikurahul commented:
This is an awesome PR. Thanks for your patience and contributions.
